### PR TITLE
MRG: Marijn fixes

### PR DESCRIPTION
**Pull request by larsoner**
Rebased version of #11.

I'll work off of this from start to finish to see if everything (also) runs for me.

### CODE REVIEW

**Comment by jasmainak** [comment]
okay great, keep me posted

**Comment by larsoner** [comment]
Okay I am still working on the source space stuff, but in the meantime this should be ready for review. Sorry about the horrible diff :(

File: scripts/processing/02-extract_events.py [hunk 1]
```diff
@@ -6,37 +6,32 @@
...
                                  consecutive='increasing', mask=mask,
-                                 mask_type='not_and', min_duration=0.003,
-                                 verbose=True)
+                                 mask_type='not_and', min_duration=0.003)
 
-        print("S %s - R %s" % (subject, run))
+        print("  S %s - R %s" % (subject, run))
 
-        fname_events = op.join(data_path, 'run_%02d_filt_sss-eve.fif' % run)
+        fname_events = op.join(out_path, 'run_%02d-eve.fif' % run)
```
Comment (larsoner) [inline scripts/processing/02-extract_events.py:33]: - events are independent of processing, process original data files

File: scripts/processing/01-anatomy.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+            mne.write_bem_solution(fname_bem, bem)
+
+    # Create the surface source space
+    fname_src = op.join(subjects_dir, subject, 'bem', '%s-%s-src.fif'
+                        % (subject, spacing))
+    if not op.isfile(fname_src):
+        print('  Setting up source space')
+        src = mne.setup_source_space(subject, spacing,
+                                     subjects_dir=subjects_dir)
+        mne.write_source_spaces(fname_src, src)
```
Comment (larsoner) [inline scripts/processing/01-anatomy.py:134]: - move anotomy-only-dependent steps (BEM, source space) to anatomy script
- only do steps that still need to be done (with note to this effect) to make it easier to restart this long process
- compute source space with CPS to improve depth weighting

File: scripts/processing/02-extract_events.py (hunk 1 above)
Comment (larsoner) [inline scripts/processing/02-extract_events.py:33]: - events are independent of processing, process original file and change name to reflect it

File: scripts/processing/03-maxwell_filtering.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+            # High-pass EOG to get reasonable thresholds in autoreject
+            picks_eog = mne.pick_types(raw.info, meg=False, eog=True)
+            raw_sss.filter(
+                l_freq, None, picks=picks_eog, l_trans_bandwidth='auto',
+                filter_length='auto', phase='zero', fir_window='hann',
+                fir_design='firwin')
+            raw_sss.save(raw_out, overwrite=True)
+
+
+run_maxwell_filter(subject_id=3)  # Only for sub003
```
Comment (larsoner) [inline scripts/processing/03-maxwell_filtering.py:138]: - lots of fixes here to make it match the MF processing as closely as possible

File: scripts/processing/03-maxwell_filtering.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        try:
+            raw = mne.io.read_raw_fif(raw_in)
+        except AttributeError:
+            # Some files on openfmri are corrupted and cannot be read.
+            warn('Could not read file %s. '
+                 'Skipping run %s from subject %s.' % (raw_in, run, subject))
+            continue
+        print('  Run %s' % (run,))
+
+        raw.set_channel_types({
```
Comment (jasmainak) [inline scripts/processing/03-maxwell_filtering.py:63]: Is this required? Maxfilter automatically detects MEG channels, no? Do the non-MEG channel types matter? In fact, I was a bit surprised to see no `picks` in `maxwell_filter` ..

File: scripts/processing/05-run_ica.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    # with blinks at the epoching stage).
+    print('  Fitting ICA')
+    ica = ICA(method='fastica', random_state=42, n_components=n_components)
+    picks = mne.pick_types(raw.info, meg=True, eeg=False, eog=False,
+                           stim=False, exclude='bads')
+    ica.fit(raw, picks=picks, reject=dict(grad=4000e-13, mag=4e-12),
+            decim=11)
+    print('  Fit %d components (explaining at least %0.1f%% of the variance)'
+          % (ica.n_components_, 100 * n_components))
+    ica.save(ica_name)
```
Comment (larsoner) [inline scripts/processing/05-run_ica.py:55]: - change structure to concat then do a single ICA

File: scripts/processing/05-run_ica.py [hunk 6]
```diff
@@ -30,19 +31,28 @@ def run_ica(subject_id, tsss=None):
...
     if tsss:
+        # SSS reduces the data rank and the noise levels, so let's include more
+        # ICA components here
+        n_components = 0.999
         ica_name = op.join(meg_dir, subject,
                            'run_concat-tsss_%d-ica.fif' % tsss)
     else:
+        n_components = 0.98
         ica_name = op.join(meg_dir, subject, 'run_concat-ica.fif')
+    # Here we only use MEG because we only eliminate ECG artifacts, which
```
Comment (jasmainak) [inline scripts/processing/05-run_ica.py]: only use MEG -> compute ICA on only MEG

File: scripts/processing/03-maxwell_filtering.py (hunk 4 above)
Comment (larsoner) [reply scripts/processing/03-maxwell_filtering.py:63]: It's required to make this otherwise equivalent to what is done in `04-python_filtering.py`, which is a good idea if we ever plan to compare the files (and we do later)

File: scripts/processing/05-run_ica.py [hunk 7]
```diff
@@ -30,19 +31,28 @@ def run_ica(subject_id, tsss=None):
...
+    # are not prevalent in EEG (blink artifats are, but we will remove trials
```
Comment (jasmainak) [inline scripts/processing/05-run_ica.py]: artifats -> artifacts

okay I stop with typos. I will try to avoid bike shedding :)

File: scripts/processing/03-maxwell_filtering.py (hunk 3 above)
Comment (jasmainak) [reply scripts/processing/03-maxwell_filtering.py:138]: awesome!

File: scripts/processing/05-run_ica.py [hunk 8]
```diff
@@ -30,19 +31,28 @@ def run_ica(subject_id, tsss=None):
         raws.append(mne.io.read_raw_fif(run_fname))
     raw = mne.concatenate_raws(raws)
     if tsss:
+        # SSS reduces the data rank and the noise levels, so let's include more
+        # ICA components here
```
Comment (jasmainak) [inline scripts/processing/05-run_ica.py]: does it reduce the number of components or is that you need a higher fraction to get the same number of components as in non-SSS data?

File: scripts/processing/06-make_epochs.py [hunk 9]
```diff
@@ -77,42 +80,50 @@ def run_epochs(subject_id, tsss=False):
...
     epochs = mne.Epochs(raw, events, events_id, tmin, tmax, proj=True,
                         picks=picks, baseline=baseline, preload=True,
                         decim=5, reject=None)
+    print('  Interpolating bad channels')
     epochs.interpolate_bads()
+    print('  Getting rejection thresholds')
     reject = get_rejection_threshold(epochs)
     epochs.drop_bad(reject=reject)
+    print('  Dropped %0.1f%% of epochs'
+          % (100 * np.mean([len(x) > 0 for x in epochs.drop_log])))
```
Comment (jasmainak) [inline scripts/processing/06-make_epochs.py]: isn't there a function in MNE which gives the same info? just want to keep the code simple as it's meant to be educational as well

File: scripts/processing/06-make_epochs.py [hunk 10]
```diff
@@ -123,6 +134,6 @@ def run_epochs(subject_id, tsss=False):
 ###############################################################################
 # Let us make the script parallel across subjects
 
-parallel, run_func, _ = parallel_func(run_epochs, n_jobs=1)
+parallel, run_func, _ = parallel_func(run_epochs, n_jobs=N_JOBS)
```
Comment (jasmainak) [inline scripts/processing/06-make_epochs.py]: wait, this does not consume humongous memory?

Comment (jasmainak) [reply scripts/processing/06-make_epochs.py]: what's the memory footprint with something like `n_jobs = 4`?

Comment (larsoner) [reply scripts/processing/06-make_epochs.py]: about 1.5 GB per job

File: scripts/processing/09-run_time_decoding.py [hunk 11]
```diff
@@ -51,7 +51,8 @@ def run_time_decoding(subject_id, condition1, condition2):
     epochs.pick_types(meg='mag').pick_channels(ch_names)
 
     # Use AUC because chance level is same regardless of the class balance
-    td = TimeDecoding(predict_mode='cross-validation', scorer='roc_auc')
+    td = TimeDecoding(predict_mode='cross-validation', scorer='roc_auc',
+                      n_jobs=N_JOBS)
```
Comment (jasmainak) [inline scripts/processing/09-run_time_decoding.py]: hmm ... why did you change this? To save memory? Perhaps worth adding a comment because it breaks the pattern we use in our other scripts

File: scripts/processing/06-make_epochs.py (hunk 10 above)
Comment (larsoner) [reply scripts/processing/06-make_epochs.py]: ... oh wait, at the writing stage it does load all epochs at once. (Too bad we can't change this, maybe we should at the MNE level if possible.) That brings it up to be quite large. I'll revert to 1.

File: scripts/processing/09-run_time_decoding.py (hunk 11 above)
Comment (larsoner) [reply scripts/processing/09-run_time_decoding.py]: Yes it saves memory and should be roughly as fast

File: scripts/processing/05-run_ica.py (hunk 8 above)
Comment (larsoner) [reply scripts/processing/05-run_ica.py]: SSS reduces rank, to recover the right number of components using ICA we need it to be this high

File: scripts/processing/06-make_epochs.py (hunk 9 above)
Comment (larsoner) [reply scripts/processing/06-make_epochs.py]: not that I know of. `epochs.selection` gives indices into the original but that won't tell you how many there originally were

Comment (jasmainak) [reply scripts/processing/06-make_epochs.py]: what about `epochs.drop_log_stats()` ?

File: scripts/processing/05-run_ica.py (hunk 8 above)
Comment (jasmainak) [reply scripts/processing/05-run_ica.py]: yes, but I guess what I was saying is that it's not clear what you mean by "more". I guess you meant more compared to what it would do for n_components = 0.98. But it could also mean: more number of components than non-SSS data

File: scripts/processing/01-anatomy.py [hunk 12]
```diff
@@ -93,19 +98,29 @@ def process_subject_anat(subject_id, force_recon_all=False):
...
-        bem_model = mne.make_bem_model(
-            subject, ico=4, subjects_dir=subjects_dir)
-        mne.write_bem_surfaces(fname_bem_surfaces, bem_model)
-    fname_bem = op.join(subjects_dir, subject, 'bem',
-                        '%s-5120-5120-5120-bem-sol.fif' % subject)
-    if not op.isfile(fname_bem):
-        print('  Computing BEM solution')
-        bem = mne.make_bem_solution(bem_model)
-        mne.write_bem_solution(fname_bem, bem)
+    for n_layers in (1, 3):
```
Comment (jasmainak) [inline scripts/processing/01-anatomy.py]: why do we need to do this?

Comment (larsoner) [reply scripts/processing/01-anatomy.py]: I was trying to get 3-layer BEMs to work. I'm optimistic that it might at some point. In the meantime it shouldn't hurt

File: scripts/results/plot_maxfilter.py [hunk 13]
```diff
@@ -11,65 +12,94 @@
...
+t = t + raw_sss_mf.first_samp / raw_sss_mf.info['sfreq']
+# The head positions in the FIF file are all zero for invalid positions
+# so let's remove them, and then concatenate our times.
+mask = (head_pos != 0).any(axis=0)
+head_pos = np.concatenate((t[np.newaxis], head_pos)).T[mask]
+# In this dataset due to old MaxFilter (2.2.10), data are uniformly
+# sampled at 1 Hz, so we save some processing time in
+# maxwell_filter by downsampling.
+skip = int(round(raw_sss_mf.info['sfreq']))
+head_pos = head_pos[::skip]
```
Comment (jasmainak) [inline scripts/results/plot_maxfilter.py]: Just wondering ... this couldn't be a function in MNE?

File: scripts/results/plot_sensor_spatio_temporal_cluster_stats.py [hunk 14]
```diff
@@ -65,10 +64,7 @@
 
 # Make a triangulation between EEG channels locations to
 # use as connectivity for cluster level stat
-# XXX : make a mne.channels.make_eeg_connectivity function
-lay = mne.channels.make_eeg_layout(contrast.info)
-neigh = spatial.Delaunay(lay.pos[:, :2]).vertices
-connectivity = mne.surface.mesh_edges(neigh)
+connectivity = mne.channels.find_ch_connectivity(contrast.info, 'eeg')[0]
```
Comment (jasmainak) [inline scripts/results/plot_sensor_spatio_temporal_cluster_stats.py:62]: interesting ...

File: scripts/results/plot_source_stats.py [hunk 15]
```diff
@@ -26,30 +26,33 @@
...
 
-stat_fun = partial(ttest_1samp_no_p, sigma=0.5)
+stat_fun = partial(ttest_1samp_no_p, sigma=0.01)
 T_obs, clusters, cluster_p_values, H0 = clu = \
     spatio_temporal_cluster_1samp_test(X, connectivity=connectivity, n_jobs=1,
                                        threshold=t_threshold,
                                        stat_fun=stat_fun, verbose=True)
 
-good_cluster_inds = np.where(cluster_p_values < p_threshold)[0]
+good_cluster_inds = np.where(cluster_p_values < 0.05)[0]
```
Comment (jasmainak) [inline scripts/results/plot_source_stats.py]: we don't use `p_threshold` anymore?

File: scripts/results/plot_maxfilter.py (hunk 13 above)
Comment (larsoner) [reply scripts/results/plot_maxfilter.py]: It could, but most users shouldn't need to do it. The right way (also in the comments) is to use `maxfilter -hp` and `read_head_pos`, but here it's inconvenient because we have the matching / incompleteness problem of the FTP site files, and even if we got them working we'd have to distribute them here.

File: scripts/results/plot_source_stats.py (hunk 15 above)
Comment (larsoner) [reply scripts/results/plot_source_stats.py]: it is used to compute `t_threshold` which is used as the *initial* thresholding during clustering

File: scripts/results/plot_maxfilter.py (hunk 13 above)
Comment (jasmainak) [reply scripts/results/plot_maxfilter.py]: Well, we could have fetched it using `urllib` but anyway doesn't the strategy your propose mean that the user would need to have an Elekta maxfilter license?

Comment (larsoner) [reply scripts/results/plot_maxfilter.py]: All strategies we currently support (the one used here and the one in comments here) for doing movement compensation currently require a MaxFilter license, yes

We have our own position estimation code but it's private and not extensively tested enough for public consumption

Comment (jasmainak) [reply scripts/results/plot_maxfilter.py]: Oh, so the estimated head positions are saved in the CHPI channels? makes sense. I guess that would be a caveat to add then to the manuscript. It's comparable up to the estimation of head position

File: scripts/results/plot_source_stats.py (hunk 15 above)
Comment (larsoner) [reply scripts/results/plot_source_stats.py]: BTW this code is untested currently because I'm still doing `recon-all`s

Comment (jasmainak) [reply scripts/results/plot_source_stats.py]: okay, before we merge this PR, could you share the html pages you built on your fork + a summary of the changes because they have to reflect in the manuscript (sort of like the what's new I guess :).

**Comment by jasmainak** [comment]
@larsoner ready to merge? The PR still says WIP

**Comment by larsoner** [comment]
Yep

**Comment by jasmainak** [comment]
Incredible. Thanks @larsoner ! Let's update the paper now

**Comment by jasmainak** [comment]
Also thanks to @wmvanvliet !

**Comment by agramfort** [comment]
:beers: !
